slice out of bounds

Using gocritic via golangci-lint:

WARN [runner] Can't run linter goanalysis_metalinter: goanalysis_metalinter: gocritic: package "main" (isInitialPkg: true, needAnalyzeSource: true): runtime error: slice bounds out of range [2:1] 
ERRO Running error: can't run linter goanalysis_metalinter
goanalysis_metalinter: gocritic: package "main" (isInitialPkg: true, needAnalyzeSource: true): runtime error: slice bounds out of range [2:1]

go get -tool github.com/go-critic/go-critic/cmd/go-critic@v0.14.1
    go tool go-critic check -enableAll ./...

still fails:

```
dupOption: error: runtime error: slice bounds out of range [2:1]
panic: runtime error: slice bounds out of range [2:1] [recovered, repanicked]

goroutine 2910 [running]:
main.(*program).checkFile.func1.1()
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:163 +0x19d       
panic({0x893200?, 0xc008b40330?})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/runtime/panic.go:783 +0x132
github.com/go-critic/go-critic/checkers.(*dupOptionChecker).getVariadicArgs(0xc220a0?, 0xc002e1ac00)
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/dupOption_checker.go:79 +0x14c       
github.com/go-critic/go-critic/checkers.(*dupOptionChecker).VisitExpr(0xc003ca43d0, {0x965cf8?, 0xc002e1ac00?})
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/dupOption_checker.go:42 +0x66       
github.com/go-critic/go-critic/checkers/internal/astwalk.(*exprWalker).WalkFile.func1({0x964238?, 0xc002e1ac00?})
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/internal/astwalk/expr_walker.go:25 +0x5b       
go/ast.inspector.Visit(0xc00975a230, {0x964238?, 0xc002e1ac00?})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:361 +0x2b
go/ast.Walk({0x962760?, 0xc00975a230?}, {0x964238, 0xc002e1ac00})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:34 +0x4c
go/ast.walkList[...](...)
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:21
go/ast.Walk({0x962760?, 0xc00975a230?}, {0x964238, 0xc002e1ac80})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:121 +0x26b4
go/ast.walkList[...](...)
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:21
go/ast.Walk({0x962760?, 0xc00975a230?}, {0x9644e0, 0xc002e1ad40})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:194 +0x1eae
go/ast.walkList[...](...)
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:21
go/ast.Walk({0x962760?, 0xc00975a230?}, {0x9640f8, 0xc0020831d0})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:211 +0x1d6e
go/ast.Walk({0x962760?, 0xc00975a230?}, {0x9647b0, 0xc002083200})
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:332 +0x6bc
go/ast.Inspect(...)
        /nix/store/x322dk8vnsfbr9p31175zhqcc9534iql-go-1.25.0/share/go/src/go/ast/walk.go:377
github.com/go-critic/go-critic/checkers/internal/astwalk.(*exprWalker).WalkFile(0xc003ca43e0, 0xc0021101e0)
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/internal/astwalk/expr_walker.go:23 +0x139       
github.com/go-critic/go-critic/linter.(*Checker).Check(...)
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/linter/linter.go:160       
main.(*program).checkFile.func1()
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:170 +0xa5       
created by main.(*program).checkFile in goroutine 1
        $GOPATH/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:151 +0xe5
```

try this code, the compile is not completed, so the real call.Args's length is 1

```go
package main

type Config struct {}
type Option func(*Config)

func NewService(name string, op Option, options ...Option) *Config {
    return nil
}

func main() {
    // This call will trigger the go-critic panic
    NewService("worker")
}
```
and
```
go-critic check -enableAll ./...
```

and it cause panic

```
dupOption: error: runtime error: slice bounds out of range [2:1]
panic: runtime error: slice bounds out of range [2:1] [recovered]
        panic: runtime error: slice bounds out of range [2:1]

goroutine 2158 [running]:
main.(*program).checkFile.func1.1()
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:163 +0x19d
panic({0x8874c0?, 0xc000d86018?})
        /usr/local/go/src/runtime/panic.go:787 +0x132
github.com/go-critic/go-critic/checkers.(*dupOptionChecker).getVariadicArgs(0xbfdf10?, 0xc000f60c40)
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/dupOption_checker.go:79 +0x150
github.com/go-critic/go-critic/checkers.(*dupOptionChecker).VisitExpr(0xc0003dc880, {0x955560?, 0xc000f60c40?})
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/dupOption_checker.go:42 +0x66
github.com/go-critic/go-critic/checkers/internal/astwalk.(*exprWalker).WalkFile.func1({0x952c38?, 0xc000f60c40?})
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/internal/astwalk/expr_walker.go:25 +0x5b
go/ast.inspector.Visit(0xc001994160, {0x952c38?, 0xc000f60c40?})
        /usr/local/go/src/go/ast/walk.go:361 +0x2b
go/ast.Walk({0x951240?, 0xc001994160?}, {0x952c38, 0xc000f60c40})
        /usr/local/go/src/go/ast/walk.go:34 +0x4c
go/ast.Walk({0x951240?, 0xc001994160?}, {0x952e68, 0xc0014e36f0})
        /usr/local/go/src/go/ast/walk.go:183 +0x3016
go/ast.walkList[...](...)
        /usr/local/go/src/go/ast/walk.go:21
go/ast.Walk({0x951240?, 0xc001994160?}, {0x952af8, 0xc00074bda0})
        /usr/local/go/src/go/ast/walk.go:211 +0x23dd
go/ast.Walk({0x951240?, 0xc001994160?}, {0x9531b0, 0xc00074be00})
        /usr/local/go/src/go/ast/walk.go:332 +0x59c
go/ast.Inspect(...)
        /usr/local/go/src/go/ast/walk.go:372
github.com/go-critic/go-critic/checkers/internal/astwalk.(*exprWalker).WalkFile(0xc0003dc890, 0xc0006fc640)
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/internal/astwalk/expr_walker.go:23 +0x139
github.com/go-critic/go-critic/linter.(*Checker).Check(...)
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/linter/linter.go:160
main.(*program).checkFile.func1()
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:170 +0xa5
created by main.(*program).checkFile in goroutine 1
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:151 +0xe5
```

```
    linttest.go:105: ~/go-critic/checkers/testdata/dupOption/positive_tests.go:149:21: not enough arguments in call to NewService
```

(can be ran with `make test-checker dupOption`)

maybe the tool cause panic first before it report compile problem

```
dupOption: error: runtime error: slice bounds out of range [2:1]
panic: runtime error: slice bounds out of range [2:1] [recovered]
        panic: runtime error: slice bounds out of range [2:1]

goroutine 2158 [running]:
main.(*program).checkFile.func1.1()
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:163 +0x19d
panic({0x8874c0?, 0xc000d86018?})
        /usr/local/go/src/runtime/panic.go:787 +0x132
github.com/go-critic/go-critic/checkers.(*dupOptionChecker).getVariadicArgs(0xbfdf10?, 0xc000f60c40)
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/dupOption_checker.go:79 +0x150
github.com/go-critic/go-critic/checkers.(*dupOptionChecker).VisitExpr(0xc0003dc880, {0x955560?, 0xc000f60c40?})
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/dupOption_checker.go:42 +0x66
github.com/go-critic/go-critic/checkers/internal/astwalk.(*exprWalker).WalkFile.func1({0x952c38?, 0xc000f60c40?})
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/internal/astwalk/expr_walker.go:25 +0x5b
go/ast.inspector.Visit(0xc001994160, {0x952c38?, 0xc000f60c40?})
        /usr/local/go/src/go/ast/walk.go:361 +0x2b
go/ast.Walk({0x951240?, 0xc001994160?}, {0x952c38, 0xc000f60c40})
        /usr/local/go/src/go/ast/walk.go:34 +0x4c
go/ast.Walk({0x951240?, 0xc001994160?}, {0x952e68, 0xc0014e36f0})
        /usr/local/go/src/go/ast/walk.go:183 +0x3016
go/ast.walkList[...](...)
        /usr/local/go/src/go/ast/walk.go:21
go/ast.Walk({0x951240?, 0xc001994160?}, {0x952af8, 0xc00074bda0})
        /usr/local/go/src/go/ast/walk.go:211 +0x23dd
go/ast.Walk({0x951240?, 0xc001994160?}, {0x9531b0, 0xc00074be00})
        /usr/local/go/src/go/ast/walk.go:332 +0x59c
go/ast.Inspect(...)
        /usr/local/go/src/go/ast/walk.go:372
github.com/go-critic/go-critic/checkers/internal/astwalk.(*exprWalker).WalkFile(0xc0003dc890, 0xc0006fc640)
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/checkers/internal/astwalk/expr_walker.go:23 +0x139
github.com/go-critic/go-critic/linter.(*Checker).Check(...)
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/linter/linter.go:160
main.(*program).checkFile.func1()
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:170 +0xa5
created by main.(*program).checkFile in goroutine 1
        /home/xxxx/go/pkg/mod/github.com/go-critic/go-critic@v0.14.1/cmd/go-critic/check.go:151 +0xe5
```

And inside golangci-lint, the compilation error is detected:

```console
 $ golangci-lint run               
main.go:1: : # github.com/golangci/sandbox
./main.go:11:13: not enough arguments in call to NewService                                                                                                                                                                           
        have (string)                                                                                                                                                                                                                 
        want (string, Option, ...Option) (typecheck)                                                                                                                                                                                  
package main
1 issues:
* typecheck: 1
```

--- a/.gitignore
+++ b/.gitignore
@@ -6,6 +6,7 @@
 /go-critic
 .idea
 pkg/
+/bin/
 
 # Test binary, build with `go test -c`
 *.test

--- a/checkers/dupOption_checker.go
+++ b/checkers/dupOption_checker.go
@@ -75,6 +75,11 @@ func (c *dupOptionChecker) getVariadicArgs(call *ast.CallExpr) ([]ast.Expr, type
 	if !ok {
 		return nil, nil
 	}
+
+	if last > len(call.Args) {
+		return nil, nil
+	}
+
 	argType := sliceType.Elem()
 	return call.Args[last:], argType
 }

--- a/checkers/testdata/dupOption/negative_tests.go
+++ b/checkers/testdata/dupOption/negative_tests.go
@@ -10,3 +10,18 @@ func doSome2(w, h int) {
 		withImage(nil, ""),
 	)
 }
+
+func f(a int, b string, c ...float64) {}
+
+func g() (int, string, float64) {
+	return 42, "hello", 3.14
+}
+
+func h() (int, string) {
+	return 42, "hello"
+}
+
+func _() {
+	f(g())
+	f(h())
+}

